### PR TITLE
agile-ticketing-system

### PR DESCRIPTION
**Pull request by Boltmerz**
Add user routes

### CODE REVIEW

File: src/app.js [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+require('./dataBase/mongoose')
+const User = require('./models/user')
+const userRouter = require('./routes/user')
+
+
+const app = express()
+app.use(express.json())
+app.use(userRouter)
+
+module.exports = app
```
Comment (Boltmerz) [inline src/app.js]: What is the purpose of this file? It appears to NOT be used.

File: src/index.js [hunk 2]
```diff
@@ -1,10 +1,11 @@
 const express = require('express')
 require('./dataBase/mongoose')
-
-const app = express()
+const User = require('./models/user')
+const userRouter = require('./routes/user')
 
 const port = process.env.PORT
```
Comment (Boltmerz) [inline src/index.js:7]: The environment variables should be fetched and set in the config file and the value should be imported and used here.

File: src/routes/user.js [hunk 3]
```diff
@@ -0,0 +1,65 @@
+const express = require('express')
+const router = new express.Router()
+const User = require('../models/user')
+const mongoose = require('mongoose')
```
Comment (Boltmerz) [inline src/routes/user.js:4]: Please add the routes validations.

File: src/routes/user.js [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+router.post('/users', async(req, res) => {
+    const user = new User(req.body)
+        try{
+            await user.save()
```
Comment (Boltmerz) [inline src/routes/user.js]: You should move the code that calls APIs/DB into a services folder. Create services/user.js and move the DB code to a functions there and call the functions inside the controller which is this file.

File: src/routes/user.js [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+     const isValidOperation = updates.every((update) => {
+         return allwoedUpdates.includes(update)
+     })
+
+     if(!isValidOperation){
+         return res.status(400).send({Error: 'Invalid Updates'})
+     }
+
+     try{
+        const user = await User.findById(req.params.id)
```
Comment (Boltmerz) [inline src/routes/user.js]: Instead of using "findById()" then "save()", please use "findByIdAndUpdate()"

File: package.json [hunk 6]
```diff
@@ -25,7 +29,12 @@
     "validator": "^13.1.17"
   },
   "devDependencies": {
+    "chai": "^4.2.0",
```
Comment (Boltmerz) [inline package.json]: Remove this package.

File: package.json [hunk 7]
```diff
@@ -25,7 +29,12 @@
...
+    "chai-http": "^4.3.0",
```
Comment (Boltmerz) [inline package.json]: Remove this package.

File: package.json [hunk 8]
```diff
@@ -25,7 +29,12 @@
...
     "env-cmd": "^8.0.2",
-    "nodemon": "^2.0.6"
+    "jest": "^26.6.2",
+    "mocha": "^8.2.1",
```
Comment (Boltmerz) [inline package.json]: Remove this package.

File: test/fixtures/db.js [hunk 9]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+
+const setupDatabase = async () => {
+    await User.deleteMany() 
+    await new User(userOne).save() 
+    await new User(userTwo).save()
+}
+
+module.exports = {
```
Comment (Boltmerz) [inline test/fixtures/db.js:30]: This is incorrect use of fixtures. Please read about fixtures in jest.

File: test/user.test.js [hunk 10]
```diff
@@ -0,0 +1,67 @@
+const { expect } = require('chai')
+const request = require('supertest')
+const app = require('../src/app')
+const User = require('../src/models/user')
+const {userOneId, userOne, setupDatabase} = require('./fixtures/db')
+
+beforeEach(setupDatabase)
```
Comment (Boltmerz) [inline test/user.test.js:7]: YOu only need to mock the controller function to return the response object that is suitable for you test.

**Review by Boltmerz — COMMENTED** [review]
Please see the comments.

**Review by KayedObeidat — APPROVED** [review]
There are my new updates, I have some problems with the id validation with the read user route and update route so I kept the API's/DB calls inside the routes.

I used the apps folder because the test folder doesn't work without it.

**Pull request closed without merging**